gflownet/envs/cube.py

```diff
@@ -897,34 +969,36 @@ def _get_logprobs_forward(
         do_increments = ~is_eos
         if torch.any(do_increments):
             # Get absolute increments
-            increments_abs = actions[do_increments]
-            # Get minimum increments
-            min_increments = torch.full_like(
-                increments_abs, self.min_incr, dtype=self.float, device=self.device
-            )
-            min_increments[is_source[do_increments]] = 0.0
-            # Get relative increments
-            increments_rel = self.absolute_to_relative_increments(
-                states_from_tensor[do_increments],
-                increments_abs,
-                min_increments,
-                self.max_val,
-                is_backward=False,
-            )
+            increments = actions[do_increments, :-1]
+            # Compute relative increments from absolute increments if state is not
+            # source
+            is_relative = ~is_source[do_increments]
+            if torch.any(is_relative):
+                states_from_rel = tfloat(
+                    states_from_tensor[do_increments],
+                    float_type=self.float,
+                    device=self.device,
+                )[is_relative]
+                increments[is_relative] = self.absolute_to_relative_increments(
+                    states_from_rel,
+                    increments[is_relative],
+                    is_backward=False,
+                )
+            # Compute diagonal of the Jacobian (see _get_jacobian_diag()) if state is
+            # not source
+            is_relative = torch.logical_and(do_increments, ~is_source)
+            if torch.any(is_relative):
+                jacobian_diag[is_relative] = self._get_jacobian_diag(
+                    states_from_rel,
+                    is_backward=False,
+                )
             # Get logprobs
             distr_increments = self._make_increments_distribution(
                 policy_outputs[do_increments]
             )
             # Clamp because increments of 0.0 or 1.0 would yield nan
             logprobs_increments_rel[do_increments] = distr_increments.log_prob(
-                torch.clamp(increments_rel, min=1e-6, max=(1 - 1e-6))
-            )
-            # Compute diagonal of the Jacobian (see _get_jacobian_diag())
-            jacobian_diag[do_increments] = self._get_jacobian_diag(
-                states_from_tensor[do_increments],
-                min_increments,
-                self.max_val,
-                is_backward=False,
+                torch.clamp(increments, min=self.epsilon, max=(1 - self.epsilon))
```

Comment:
I would actually do an `assert not torch.any(torch.isinf(increments))` prior to all clampings just to make sure EOS actions don't slip by as before with the backward sampling bug.

Reply:
Ok, I am making it a bit more aggressive with
```
assert torch.any(torch.isfinite(increments))
```

Reply:
https://github.com/alexhernandezgarcia/gflownet/pull/214/commits/3f38836052f02266ec6746f850ae857d005b6eaa